### PR TITLE
Native tpms context

### PR DESCRIPTION
**Pull request by Superhepper**
In this pull request the `TpmsContext` struct that was residing inside `utils` module have been moved to the `structures` module and been renamed to `SavedTpmContext`.

### CODE REVIEW

File: tss-esapi/src/constants/tss.rs [hunk 1]
```diff
@@ -644,6 +644,10 @@ pub const TSS2_RESMGR_TPM_RC_LAYER: TSS2_RC = 0x000C0000; /* base is a TPM2_RC_*
 
 pub const TSS2_RC_SUCCESS: TSS2_RC = 0x00000000;
 
+pub const TPMI_DH_SAVED_TRANSIENT: TPMI_DH_SAVED = 0x80000000; /* an ordinary transient object */
```
Comment (wiktor-k) [inline tss-esapi/src/constants/tss.rs:647]: It's too bad that these comments are not doc comments above the constants but at least the change is consistent :sweat_smile:

File: tss-esapi/src/interface_types/data_handles.rs [hunk 2]
```diff
@@ -1,10 +1,17 @@
-// Copyright 2020 Contributors to the Parsec project.
```
Comment (wiktor-k) [inline tss-esapi/src/interface_types/data_handles.rs:1]: Copyright header got removed?

File: tss-esapi/src/interface_types/data_handles.rs [hunk 3]
```diff
@@ -1,10 +1,17 @@
...
+use crate::{
+    handles::{
+        HmacSessionTpmHandle, NvIndexTpmHandle, PcrTpmHandle, PersistentTpmHandle,
+        PolicySessionTpmHandle, TpmHandle, TransientTpmHandle,
+    },
+    tss2_esys::{TPMI_DH_CONTEXT, TPMI_DH_SAVED},
+    Error, Result, WrapperErrorKind,
+};
+use std::convert::TryFrom;
+/// Enum representing the 'Object' data handles interface type.
```
Comment (wiktor-k) [inline tss-esapi/src/interface_types/data_handles.rs:11]: I think a newline here would be :+1:

File: tss-esapi/src/interface_types/data_handles.rs [hunk 4]
```diff
@@ -53,10 +59,118 @@ pub enum Entity {
...
 #[derive(Debug, Copy, Clone)]
 pub enum Pcr {
     Pcr(PcrTpmHandle),
 }
+
+/// Enum representing the 'Context' data handles interface type.
+///
+/// # Details
+/// This corresponds to the TPMI_DH_CONTEXT interface type.
+/// This corresponds to the TPMI_DH_CONTEXT interface type. This only
```
Comment (wiktor-k) [inline tss-esapi/src/interface_types/data_handles.rs:74]: The sentence seems to be repeated twice here.

Comment (wiktor-k) [reply tss-esapi/src/interface_types/data_handles.rs:74]: I'd also use backticks in more places: `TPMI_DH_CONTEXT` and maybe "the `Context` data"

**Review by wiktor-k — DISMISSED** [review]
It looks good :+1: I've left quite a few nitpicks :sweat_smile:

File: tss-esapi/src/constants/tss.rs (hunk 1 above)
Comment (Superhepper) [reply tss-esapi/src/constants/tss.rs:647]: I mean we could add the same comment to the Enum item if that would make things better.

File: tss-esapi/src/interface_types/data_handles.rs (hunk 3 above)
Comment (Superhepper) [reply tss-esapi/src/interface_types/data_handles.rs:14]: Fixed

File: tss-esapi/src/interface_types/data_handles.rs (hunk 4 above)
Comment (Superhepper) [reply tss-esapi/src/interface_types/data_handles.rs:76]: Fixed

File: tss-esapi/src/structures/tpm_context.rs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    /// Deserialize the [SavedTpmContext] data from it's bytes representation of the TCG
+    /// TPMS_CONTEXT structure.
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes = <Vec<u8>>::deserialize(deserializer)?;
+        Self::unmarshall(&bytes).map_err(serde::de::Error::custom)
+    }
+}
```
Comment (ionut-arm) [inline tss-esapi/src/structures/tpm_context.rs:104]: Thanks! 🙏🏻

**Review by ionut-arm — APPROVED** [review]
Big thank you from me!

**Review by wiktor-k — APPROVED** [review]
LGTM 👌

**Pull request merged**